{% extends 'base.html' %}
{% load static %}

{% block title %}PSINDB{% endblock %}

{% block sidenav %}
<div id="sidenav-manual" class="sidenav">
    <a class="manual-main-category" href="#construction">Construction of the database</a>
        <a class="manual-sub-category" href="#construction_1">Selection of postsynaptic (PS) proteins</a>
        <a class="manual-sub-category" href="#construction_2">Comprehensive collection of PS protein-protein interactions (PPIs)</a>
    
    <a class="manual-main-category" href="#navigation">Navigating on the PSINDB database</a>
        <a class="manual-sub-category" href="#search-browse">Search/Browse</a>
        <a class="manual-sub-category" href="#entry">Entry</a>
            <a class="manual-sub-sub-category" href="#entry-function">Function</a>
            <a class="manual-sub-sub-category" href="#entry-features">Protein features</a>
            <a class="manual-sub-sub-category" href="#entry-partners-interacting-regions">Partners with known interacting regions</a>
            <a class="manual-sub-sub-category" href="#network">Network</a>
            <a class="manual-sub-sub-category" href="#isoforms">Isoforms</a>
            <a class="manual-sub-sub-category" href="#disease">Disease-causing germline mutations</a>
            <a class="manual-sub-sub-category" href="#linear-motifs">Linear motifs</a>
            <a class="manual-sub-sub-category" href="#fingerprint">Fingerprint</a>
            <a class="manual-sub-sub-category" href="#all-partners">All partners</a>
        <a class="manual-sub-category" href="#interaction-page">Interaction page</a>

    <a class="manual-main-category"href="#download-page">Download</a>
</div>
{% endblock %}

{% block content %}
    <h1 id="construction" class="lonely">Server manual</h1>

{#    <div class="container">#}
{#        <div class="row">#}
{#            <div class="col-sm">#}

                <div class="manual-content">
                    <h2 id="construction_1" >Selection of postsynaptic (PS) proteins</h2>
                    <p>
                        We downloaded the human reference proteome from <a href="https://www.uniprot.org/">UniProt</a>
                        then used four resources to define the postsynaptic proteome:
                        <a href="http://metamoodics.org/SynaptomeDB/index.php">SynaptomeDB</a>,
                        <a href="https://www.genes2cognition.org/">G2C</a>,
                        <a href="https://syngoportal.org/">SynGO</a> and
                        <a href="http://geneontology.org/">Gene ontology annotations</a>.
                        For all proteins listed in any of these resources we collected interactions from various sources.
                    </p>
                    <p>
                        Furthermore, we also display protein expression data from
                        Human Protein Atlas [<a href="https://www.proteinatlas.org">https://www.proteinatlas.org</a>],
                        however we are not using it as a filter. Proteins showing neuronal expression in Cerebral
                        cortex, Caudate or in the Hippocampus are displayed.
                    </p>
                </div>

                <div class="manual-content">
                    <h2 id="construction_2">A comprehensive collection of PS protein-protein interactions (PPIs)</h2>
                    <p>
                        We collected nearly 2000 experiments from the literature describing PPIs using at
                        <a href="https://pubmed.ncbi.nlm.nih.gov/17687370/">MIMIx</a> or
                        <a href="https://pubmed.ncbi.nlm.nih.gov/22453911/">higher</a>
                        standards. During the collection we aimed to look for papers that were not included in other
                        similar resources. To have a complete collection of PS Interactome we also display
                        interactions from IntAct <a href="https://www.ebi.ac.uk/intact/">IntAct</a>
                        and <a href="https://thebiogrid.org/">BIOGRID</a> databases.
                        In addition, we scanned the <a href="https://www.rcsb.org/">PDB database</a>
                        for complexes that have interacting residues. Furthermore, computational evidence from the
                        <a href="https://string-db.org/">STRING</a> database is also displayed
                        (above 0.8 cutoff score).
                    </p>
                    <p>Users can also search for the interaction of any protein.
                        The result page will help the user select proteins that localize to PS.</p>
                    <p>
                        PSINDB displays the interaction between human PS proteins. We considered evidence from mouse
                        and rat experiments too â€“ in this case the interaction is always mapped back the
                        orthologous human protein, using
                        <a href="https://omabrowser.org/oma/home/">OMA</a>
                        orthologous database and gene names.
                    </p>
                </div>


                <div class="manual-content">
                    <h2 id="navigation">Navigating on the PSINDB database</h2>
                    <p>
                        Data on the PSINDB database can be viewed using three distinct layouts, each displaying more
                        specific information about the PS Interactome. The first level is the <strong>search/browse page</strong>,
                        where proteins of preconstructed sets can be viewed (or any other search can be performed).
                        The next level is the <strong>protein page</strong>, where general information is displayed about the protein,
                        along with a summary of their interacting partners. Finally, the <strong>interaction page</strong> displays every
                        specific experimental (or computational) information about the selected interaction.
                    </p>
                </div>

                <div class="manual-content2">
                    <h3 id="search-browse">Search/browse page</h3>
                    <p>
                        Users can search for proteins using UniProt identifier, UniProt AC, GeneID, Gene name,
                        Ensemble Protein identifier and HGNC codes using the search bar at the top of the page.
                        After searching the results will be listed, along with the number of interactions and the
                        evidence that the protein of choice localizes into the PS.
                    </p>
                    <p>
                        Users can also search for the interaction of any protein.
                        The result page will help the user select proteins that localize to PS.
                    </p>
                    <p>
                        We also prepared sets of proteins, e.g, proteins are known to be going through phase separation,
                        proteins with PPIs having literature evidence or transmembrane proteins and more. These sets
                        can be accessed through the browse menu.
                    </p>
                    <img class="manual-image" src="{% static 'images/IMAGE1.gif' %}" alt="PSINDB logo">
                </div>

                <div class="manual-content2">
                    <h3 id="entry">Entry page</h3>
                </div>

                <div class="manual-content3">
                    <h4 id="entry-function">Function</h4>
                    <p>
                        For each protein basic information about the protein is displayed using
                        <a href="https://www.uniprot.org/">UniProt</a> functional annotations.
                    </p>
                </div>

                <div class="manual-content3">
                    <h4 id="entry-features">Protein features</h4>
                    <p>
                        Protein features display structural and functional information about the protein mirrored to
                        the protein sequence. The legend below the Features guides the user on how to interpret this information.
                        Using the scroll users can zoom into regions. Clicking into a region, the segment is highlighted
                        along with all other features.
                    </p>
                    <p>
                        <i>Topology:</i> transmembrane segments were predicted using
                        <a href="http://cctop.enzim.ttk.mta.hu/">CCTOP</a>.
                        Since CCTOP also includes experimental evidence, the corresponding entry is linked to the
                        <a href="http://htp.enzim.hu/">Human Transmembrane Proteome</a> database.
                    </p>
                    <p>
                        <i>Phase separation:</i> displays protein regions going through phase separation from
                        <a href="https://phasepro.elte.hu/">PhaSePro</a> database. Although currently the number of
                        annotated proteins is low, we expect a rapid increase of annotated data as the field emerges.
                    </p>
                    <p>
                        <i>Linear motifs:</i> short linear motifs are compact protein regions usually between 3-10 residues
                        playing crucial roles in cell regulation. The <a href="http://elm.eu.org/">ELM</a> database is
                        the most comprehensive collection of experimentally verified motifs. More information about
                        these segments can be found below at the Linear motif section of the entry page.
                    </p>
                    <p>
                        <i>Phosphorylation:</i> phosphorylations are post-translational modifications that can have a diverse
                        range of functions. They often regulate PPIs by working as switches, changing between the
                        non-modified (OFF) state and the modified (charged) (ON) state. Phosphorylation annotation were
                        collected from <a href="https://www.uniprot.org/">UniProt</a>.
                    </p>
                    <p>
                        <i>PFAM:</i> <a href="http://pfam.xfam.org/">PFAM</a> is a collection and classification of
                        protein families and domains.

                    </p>
                    <p>
                        <i>Coiled coil:</i> Coiled-coils are oligomeric helical structural units in proteins connected to a
                        wide range of functions. They can bridge large distances and connect proteins at different
                        sides of large supramolecular structures. Coiled coil segments were predicted using
                        <a href="https://toolkit.tuebingen.mpg.de/tools/deepcoil">DeepCoil</a>.
                    </p>
                    <p>
                        <i>Anchor:</i> Intrinsically disordered regions often bind to ordered globular domains.
                        Such disordered binding segments are predicted using <a href="https://iupred2a.elte.hu/">Anchor</a>.
                    </p>
                    <p>
                        <i>Disordered regions:</i> Intrinsically disordered regions (IDRs) are flexible protein regions that
                        can serve various functions. IDRs can mediate binary interactions of proteins or promote their
                        organization of large assemblies. Disordered segments were predicted using
                        <a href="https://iupred2a.elte.hu/">IUPred</a>.
                    </p>
                    <p>
                        <i>Interacting regions:</i> In this panel we mirrored the interaction regions responsible for binding
                        the proteinâ€™s partners. Experiments provide different levels of information about the
                        interacting regions. If two or more experiments overlap, we always display the information from
                        the most specific one.
                    </p>
                    <p>
                        <i>Binding-associated region (source: Intact, PSINDB):</i> a region of a molecule or a component of a
                        complex identified as being involved in an interaction. This may or may not be a region of the
                        molecule in direct contact with the interacting partner.

                    </p>
                    <p>
                        <i>Sufficient binding region (Intact, PSINDB, PDB):</i> binding will occur when this sequence range is
                        present within a molecule or part of a molecule. This region will contain the direct binding
                        region but may be longer.

                    </p>
                    <p>
                        <i>Necessary binding region (Intact, PSINDB):</i> a sequence range within a molecule identified as
                        being absolutely required for an interaction. The sequence may or may not be in direct physical
                        contact with the interaction partner.

                    </p>
                    <p>
                        <i>Atomic contact (PDB):</i> atomic contact between residues. After reconstructing the full PDB
                        structures using the biomatrix information we used
                        <a href="https://github.com/kliment-olechnovic/voronota">Voronota</a>
                        to detect residues establishing interchain interactions.
                    </p>
                    <p>
                        For experiments on orthologous proteins even if the binding segment was defined, we did not
                        map it back to the original sequence â€“ sometimes this task is trivial, however there are several
                        cases when the alignment
                        is ambiguous, especially when there are disordered segments involved.
                    </p>
                    <img class="manual-image" src="{% static 'images/IMAGE2.gif' %}" alt="PSINDB logo">
                </div>

                <div class="manual-content3">
                    <h4 id="entry-partners-interacting-regions">Partners with known interacting regions</h4>
                    <p>
                        This section displays all partners where the interacting region of at least one of the partners
                        was narrowed from the full length sequence. This annotation has the same levels as described
                        earlier in the Protein features/Interacting regions section. The link of the partner navigates
                        to the interaction page with detailed information about the experiments.
                    </p>
                    <img class="manual-image" src="{% static 'images/IMAGE3.gif' %}" alt="PSINDB logo">
                </div>

                <div class="manual-content3">

                    <h4 id="network">Network</h4>
                    <p>
                        The Network view displays interaction partners and also shows how these partners interact with
                        each other. The PS exhibits an extremely high number of PPIs, therefore we reduced the number
                        of displayed partners to 50 in this view (users can still download the full network from the
                        link below the panel). For this purpose we scored each interaction using the following method:
                        for each evidence we assigned a score. The score of any interaction is the sum of individual
                        pieces of evidence.
                    </p>
                    <table>
                        <thead>
                        <th></th>
                        <th>Experiment on human proteins</th>
                        <th>Experiment on orthologous proteins</th>
                        </thead>
                        <tbody>
                        <tr>
                            <td>Low throughput</td>
                            <td>1</td>
                            <td>0.8</td>
                        </tr>
                        <tr>
                            <td>High throughput</td>
                            <td>0.8</td>
                            <td>0.64</td>
                        </tr>
                        <tr>
                            <td>Computational</td>
                            <td>0.6</td>
                            <td>x</td>
                        </tr>
                        </tbody>
                    </table>
                    <p>
                        For the definition of low throughput, high throughput and computational evidence see the
                        "All partners" section.
                    </p>
                    <img class="manual-image" src="{% static 'images/IMAGE4.gif' %}" alt="PSINDB logo">
                </div>

                <div class="manual-content3">

                    <h4 id="isoforms">Isoforms</h4>
                    <p>
                        This section displays all isoforms, together with the interacting regions of the canonical form,
                        thus isoforms with missing binding regions can be easily captured. Isoforms were collected from
                        <a href="https://www.uniprot.org/">UniProt</a> and they were aligned using
                        <a href="https://www.ebi.ac.uk/Tools/msa/clustalo/">ClustalOmega</a>.
                    </p>
                    <img class="manual-image" src="{% static 'images/IMAGE5.gif' %}" alt="PSINDB logo">
                </div>

                <div class="manual-content3">
                    <h4 id="disease">Disease-causing germline mutations</h4>
                    <p>
                        Disease-causing germline mutations were collected from the
                        <a href="https://www.omim.org/">OMIM</a> database. We also highlight mutations that fall into
                        the interacting region of the protein and display partners binding through that region.
                    </p>
                </div>


                <div class="manual-content3">
                    <h4 id="linear-motifs">Linear motifs</h4>
                    <p>
                        Short linear motifs are compact protein regions usually between 3-10 residues playing crucial
                        roles in cell regulation. The <a href="http://elm.eu.org/">ELM</a> database is the most
                        comprehensive collection of experimentally verified motifs. We also highlight SLiMs that fall
                        into the interacting region of the protein and display partners binding through that region.
                    </p>
                    <img class="manual-image" src="{% static 'images/IMAGE6.gif' %}" alt="PSINDB logo">
                </div>

                <div class="manual-content3">
                    <h4 id="fingerprint">Fingerprint</h4>
                    <p>
                        Fingerprint provides a collection of ontology terms re-occurring in the protein or its
                        interaction partners. We collected all terms annotated for the proteins and also mapped
                        ancestor terms for the given protein. The annotation of different proteins may vary and thus
                        any statistics would be biased, therefore we only display the percentage of different
                        terms in the network and do not calculate further statistics. We also highlight terms that are
                        associated with the entry.
                    </p>
                    <p>
                        Hits can be filtered using how specific is the given term. Currently PSINDB displays three ontologies:
                        <a href="http://geneontology.org/">GO</a>: biological process,
                        <a href="http://geneontology.org/">GO</a>: molecular function
                        and <a href="https://disease-ontology.org/">DOID</a>: disease ontology.

                    </p>
                    <img class="manual-image" src="{% static 'images/Fingerprint.png' %}" alt="PSINDB logo">
                </div>

                <div class="manual-content3">
                    <h4 id="all-partners">All partners</h4>
                    <p>
                        This panel displays all interacting partners, including the ones that not localize to PS
                        according to our current knowledge. PS partners are highlighted.
                    </p>
                    <p>
                        For each partner there are two links: [view entry] navigate to the entry page of the selected
                        protein. [view interaction] navigates to the interaction page with detailed information about
                        the experiments.
                    </p>
                    <p>
                        Highest evidence display the highest level of information about the particular interaction:

                    </p>
                    <p>
                        *Low throughput (LTP): in case of PSINDB, the curator can make the call to classify the given
                        interaction. In the case of Intact and BIOGRID, we followed the following procedure: in any paper,
                        if using the same experimental setup less than ten experimental data was produced, we assigned
                        it as low throughput. Interactions from PDB are always classified as low throughput.
                    </p>
                    <p>
                        *High throughput (HTP): in case of PSINDB, the curator can make the call to classify the given
                        interaction. In the case of Intact and BIOGRID, we followed the following procedure: in any paper,
                        if using the same experimental setup ten or more experimental data was produced, we assigned it
                        as high throughput.
                    </p>
                    <p>
                        Computational: Interactions collected from the STRING database belong to this category.
                    </p>
                    <p>
                        *We are aware that this classification may have many pitfalls, however this seemed the most
                        efficient way to filter out primary screens. In several cases these interactions are further
                        validated in the same paper, therefore they will get the LTP tag.
                    </p>
                    <img class="manual-image" src="{% static 'images/IMAGE8.gif' %}" alt="PSINDB logo">
                </div>

                <div class="manual-content2">
                    <h3 id="interaction-page">Interaction page</h3>
                    <p>
                        This page contains all evidence of interaction for a pair of proteins. Information is stored
                        according to the PSI MI controlled vocabularies and can be inspected on the
                        <a href="https://www.ebi.ac.uk/ols/ontologies/mi">Ontology Lookup Services</a>.
                    </p>
                    <img class="manual-image" src="{% static 'images/IMAGE9.gif' %}" alt="PSINDB logo">
                </div>

                <div class="manual-content2">
                    <h3 id="download-page">Download</h3>
                    <p>
                        Data from PSINDB can be downloaded in the following formats:
                    </p>
                    <ul>
                        <li>experimental_mitab25.tsv  - MITAB 2.5 format file for all experimental data</li>
                        <li>PSINDB_mitab25.tsv - MITAB 2.5 format file containing data collected by our curation team</li>
                        <li>merged_binding.txt - this file contains the interacting regions, whenever it is available.
                            For interactions with multiple experiments and different regions the union of the interacting
                            region is shown
                        </li>
                        <li>postsynaptic_proteins.tsv - List of postsynaptic proteins</li>
                        <li>postsynaptic_network.tsv - Full PS network with scores</li>
                    </ul>
                </div>
{% endblock %}